lib/Conversion/TorchToTosa/TosaLegalizeCommon.cpp

```diff
@@ -27,6 +28,370 @@
+
+  //  ND: indices.shape[-1]
+  ND = indices_type.getShape()[indices_rank - 1]; // 3
+
+  if (ND > params_rank) {
+    (void)rewriter.notifyMatchFailure(
+        op, "size of last dimension of indices must be <= params rank");
+    return llvm::None;
+  }
+
+  // Calculate N, K, W, C.  (N is always 1)
+  // number of indices in each batch. product(indices.shape[0:-1]) (all but the
+  // last dimension) W = 1*4*2 = 8
+  for (int i = 0; i < (indices_rank - 1); i++) {
+    W *= indices_type.getShape()[i];
+  }
+
+  // K: range of each index, total number of inputs(chould be gather) after
+  // flattened k = 1*1*4*3 = 12
+  for (int i = 0; i < ND; i++) {
+    K *= params_type.getShape()[i];
+  }
+
+  // C: number of channels for each index : numbers of values inside each
+  // input(chould be gather) C = product(params.shape[ND:] ND = 3, params_rank,
+  // C = 1
+  for (int i = ND; i < params_rank; i++) {
+    C *= params_type.getShape()[i];
+  }
+
+  // int N = 1, W = 8, K = 12, C = 1, ND = 3;
+  SmallVector<int64_t, 3> tosa_values_shape({N, K, C});  // {1,12,1}
+  SmallVector<int64_t, 2> tosa_indices_shape({N, W});    // {1,8}
+  SmallVector<int64_t, 2> indices_matrix_shape({W, ND}); // {8,3}
+  SmallVector<int64_t, 2> indices_matrix_reducesum_shape(
+      {W, 1}); // {8,1} This is different from tf tosa code
+  SmallVector<int64_t, 3> tosa_gather_result_shape({N, W, C}); // {1,8,1}
+
+  // %2 = "tosa.reshape"(%0) {new_shape = [1, 12, 1]} : (tensor<1x4x3xf32>) ->
+  // tensor<1x12x1xf32>
+  auto tosa_values_reshape_op = tosa::CreateOpAndInfer<tosa::ReshapeOp>(
+      rewriter, op->getLoc(),
+      GetTypeFromTensorShape(tosa_values_shape, params_type.getElementType()),
+      params_value, rewriter.getI64ArrayAttr(tosa_values_shape));
+
+  // %3 = "tosa.reshape"(%1) {new_shape = [8, 3]} : (tensor<1x4x2x3xi32>) ->
+  // tensor<8x3xi32> Flatten the input indices tensor to an [W, ND] matrix.
+  auto indices_matrix_reshape_op = tosa::CreateOpAndInfer<tosa::ReshapeOp>(
+      rewriter, op->getLoc(),
+      GetTypeFromTensorShape(indices_matrix_shape,
+                             indices_type.getElementType()),
+      indices_value, rewriter.getI64ArrayAttr(indices_matrix_shape));
+
+  SmallVector<int32_t> flattened_coeff_vec; //  [12,3,1]
+  // flattened_coeff_vec = [4,3,1]
+  for (int i = 1; i < ND; i++) {
+    flattened_coeff_vec.push_back(params_type.getShape()[i]);
+  }
+  flattened_coeff_vec.push_back(1);
+
+  // flattened_coeff_vec = [12,3,1]
+  for (int i = ND - 1; i > 0; i--) {
+    flattened_coeff_vec[i - 1] *= flattened_coeff_vec[i];
+  }
+
+  // Create the tosaConstTensor for the flattened_coeff_vec
+  // %4 = "tosa.const"() {value = dense<[12, 3, 1]> : tensor<3xi32>} : () ->
+  // tensor<3xi32>
+  llvm::Optional<Value> flattened_coeff_value = getConstTensor<int32_t>(
+      rewriter, op, flattened_coeff_vec,
+      {static_cast<int64_t>(flattened_coeff_vec.size())});
+
+  if (!flattened_coeff_value)
+    return llvm::None;
```

Comment:
Same here, as well as in all other places.

Reply:
Done